### PR TITLE
Fix ForClause PositionalVar empty sequence handling

### PR DESCRIPTION
**Pull request by Robin6939**
### Description:  
Resolved an issue where ForClause's PositionalVar incorrectly returned values when faced with an empty sequence while allowing emptiness. This adjustment ensures accurate handling of edge cases in sequence evaluations, aligning behavior with expected standards.

### Reference: 
https://github.com/eXist-db/exist/issues/5235

### Tests: 
Added tests which verifies the same

### CODE REVIEW

**Review by line-o — COMMENTED** [review]
This looks really great, thank you for your contribution! What a 🎁 
Would it be OK for me to fix the formatting and commit message format or do you want to do it?

**Comment by Robin6939** [comment]
Hi @line-o,
Thank you for your review 😄 
I tried to fix the commit message. Can you help me fix the formatting. Since this is my first PR here, I am not familiar with it.

**Comment by adamretter** [comment]
@Robin6939 Wonderful to see that you made some progress after our call yesterday... Quick work!

Regards formatting, IntelliJ has a Java code formatter on the `Code` menu item, there is `Reformat Code`, but if the existing file is not correctly formatted, then this will cause more noise than you need. So I will comment directly on the formatting of your code separately...

File: exist-core/src/main/java/org/exist/xquery/ForExpr.java [hunk 1]
```diff
@@ -229,7 +229,10 @@ private void processItem(LocalVariable var, Item contextItem, Sequence in, Seque
         context.proceed(this);
         context.setContextSequencePosition(p, in);
         if (positionalVariable != null) {
-            at.setValue(new IntegerValue(this, p + 1));
+            if(contextItem.getType() == Type.ANY_ATOMIC_TYPE)
```
Comment (adamretter) [inline exist-core/src/main/java/org/exist/xquery/ForExpr.java:232]: @Robin6939 For formatting:

1. Insert a space between `if` and `(`
2. Add braces even for one line if statements.

e.g.:

```java
if (contextItem.getType() == Type.ANY_ATOMIC_TYPE) {
    at.setValue(new IntegerValue(this, 0));
} else {
    at.setValue(new IntegerValue(this, p + 1));
}
```

**Review by adamretter — CHANGES_REQUESTED** [review]
Small formatting issue.

**Review by reinhapa — APPROVED** [review]

File: exist-core/src/main/java/org/exist/xquery/ForExpr.java (hunk 1 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ForExpr.java]: how about 
```java
final int nextPosition = contextItem.getType() == Type.ANY_ATOMIC_TYPE ? 0 : p + 1;
at.setValue(new IntegerValue(this, nextPosition));
```

File: exist-core/src/main/java/org/exist/xquery/ForExpr.java [hunk 2]
```diff
@@ -229,7 +229,11 @@ private void processItem(LocalVariable var, Item contextItem, Sequence in, Seque
         context.proceed(this);
         context.setContextSequencePosition(p, in);
         if (positionalVariable != null) {
-            at.setValue(new IntegerValue(this, p + 1));
+            if (contextItem.getType() == Type.ANY_ATOMIC_TYPE) {
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/xquery/ForExpr.java:232]: how about
```
final int position = contextItem.getType() == Type.ANY_ATOMIC_TYPE ? 0 : p + 1;
at.setValue(new IntegerValue(position));
```

File: exist-core/src/test/xquery/xquery3/flwor.xql [hunk 3]
```diff
@@ -156,3 +156,17 @@ function flwor:allowing-empty($n as xs:integer) {
     return concat("[", $x, "]")
 };
 
+declare
```
Comment (line-o) [inline exist-core/src/test/xquery/xquery3/flwor.xql:159]: Do we have a test that shows the correct behaviour if `allowing empty` is not set?

File: exist-core/src/main/java/org/exist/xquery/ForExpr.java (hunk 2 above)
Comment (line-o) [reply exist-core/src/main/java/org/exist/xquery/ForExpr.java:232]: I am also wondering, if there is a better way to check for this case.
Since we want to avoid that the position is set to 0 in any other case.
Since the function `processItem` is called with contextItem set to `AtomicValue.EMPTY_VALUE` if and only if `allowing empty` is set _and_ there are no items in the binding sequence we should best check for that.
What do you think?

**Review by line-o — COMMENTED** [review]
I added a few suggestions and questions.

File: exist-core/src/test/xquery/xquery3/flwor.xql (hunk 3 above)
Comment (Robin6939) [reply exist-core/src/test/xquery/xquery3/flwor.xql:159]: There isn't any right now
Should I add them?

File: exist-core/src/main/java/org/exist/xquery/ForExpr.java (hunk 2 above)
Comment (Robin6939) [reply exist-core/src/main/java/org/exist/xquery/ForExpr.java:232]: Yeah that is better
Will make that change

**Review by dizzzz — APPROVED** [review]

**Review by line-o — APPROVED** [review]
